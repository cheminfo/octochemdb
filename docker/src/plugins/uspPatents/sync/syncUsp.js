import md5 from 'md5';

import debugLibrary from '../../../utils/Debug.js';

import { getFilesToImportForUsp } from './utils/getFilesToImportForUsp.js';
import { syncAllUspFolders } from './utils/http/syncAllUspFolders.js';
import { importUspFiles } from './utils/importUspFiles.js';
/**
 * @description performs the importation of usp patents
 * @param {*} connection - mongo connection
 * @returns {Promise} usp collection
 */
export async function sync(connection) {
  const debug = debugLibrary('syncUsp');
  try {
    // get progress
    const progress = await connection.getProgress('uspPatents');
    let allFiles;
    // get all files to import
    if (process.env.NODE_ENV === 'test') {
      allFiles = [
        {
          name: '2001.xml',
          path: `${process.env.USP_TEST_SOURCE}2001.xml`,
        },
        {
          name: '2005.xml',
          path: `${process.env.USP_TEST_SOURCE}2005.xml`,
        },
        {
          name: '2006.xml',
          path: `${process.env.USP_TEST_SOURCE}2006.xml`,
        },
        {
          name: '2007.xml',
          path: `${process.env.USP_TEST_SOURCE}2007.xml`,
        },
        {
          name: '2023.xml',
          path: `${process.env.USP_TEST_SOURCE}2023.xml`,
        },
      ];
    } else {
      allFiles = await syncAllUspFolders(connection);
    }
    const { files, lastDocument } = await getFilesToImportForUsp(
      connection,
      progress,
      allFiles,
    );
    // put file.path in array
    let sources = [];
    files.forEach((file) => {
      sources.push(file.path.replace(`${process.env.ORIGINAL_DATA_PATH}`, ''));
    });

    let shouldUpdate = false;
    if (
      progress.dateEnd !== 0 &&
      Date.now() - Number(progress.dateEnd) >
        Number(process.env.USP_PATENTS_UPDATE_INTERVAL) * 24 * 60 * 60 * 1000 &&
      md5(JSON.stringify(sources)) !== progress.sources
    ) {
      progress.dateStart = Date.now();
      shouldUpdate = true;
      await connection.setProgress(progress);
    }
    const logs = await connection.getImportationLog({
      collectionName: 'uspPatents',
      sources,
      startSequenceID: progress.seq,
    });
    if (
      (JSON.stringify(sources) !== progress.sources && shouldUpdate) ||
      progress.state !== 'updated'
    ) {
      //set progress to updating
      progress.state = 'updating';
      await connection.setProgress(progress);
      await importUspFiles(connection, progress, files, { lastDocument });
      // set progress to updated
      progress.state = 'updated';
      await connection.setProgress(progress);
      // create indexes
      const collection = await connection.getCollection('uspPatents');
      await collection.createIndex({ 'data.title': 1 });
      await collection.createIndex({ 'data.patentNumber': 1 });
      // create text index where title has more weight than abstract
      await collection.createIndex(
        { 'data.title': 'text', 'data.abstract': 'text' },
        { weights: { 'data.title': 1, 'data.abstract': 0.1 } },
      );
      await collection.createIndex({ _seq: 1 });
      progress.sources = md5(JSON.stringify(sources));
      progress.state = 'updated';
      progress.dateEnd = Date.now();
      await connection.setProgress(progress);

      logs.dateEnd = Date.now();
      logs.endSequenceID = progress.seq;
      logs.status = 'updated';

      await connection.updateImportationLog(logs);
    }
  } catch (e) {
    if (connection) {
      debug(e.message, {
        collection: 'uspPatents',
        connection,
        stack: e.stack,
      });
    }
  }
}
