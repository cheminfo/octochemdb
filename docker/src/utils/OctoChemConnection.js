import delay from 'delay';
import dotenv from 'dotenv';
import md5 from 'md5';
import { MongoClient } from 'mongodb';

import debugLibrary from './Debug.js';

dotenv.config();
const debug = debugLibrary('OctoChemConnection');

export function OctoChemConnection() {
  this.client = new MongoClient(process.env.MONGODB_URL, {
    keepAlive: true,
    connectTimeoutMS: 6 * 60 * 60 * 1000,
    socketTimeoutMS: 6 * 60 * 60 * 1000,
  });
}

OctoChemConnection.prototype.close = async function close() {
  if (this.connection) return this.connection.close();
  return undefined;
};

OctoChemConnection.prototype.getCollectionNames =
  async function getCollectionNames() {
    return (await (await this.getDatabase()).listCollections().toArray()).map(
      (entry) => entry.name,
    );
  };

OctoChemConnection.prototype.getCollection = async function getCollection(
  collectionName,
) {
  return (await this.getDatabase()).collection(collectionName);
};

OctoChemConnection.prototype.getImportationLog =
  async function getImportationLog(options) {
    const { collectionName, sources, startSequenceID } = options;
    const logsCollection = await this.getImportationLogsCollection();
    const sourcesHash = md5(JSON.stringify(sources));
    const _id = sourcesHash;
    let logs = await logsCollection.find({ _id }).next();
    if (logs === null) {
      logs = {
        _id,
        collectionName,
        sources,
        sourcesHash,
        dateStart: Date.now(),
        dateEnd: Date.now(),
        startSequenceID,
        endSequenceID: 0,
        status: 'updating',
      };
      await logsCollection.insertOne(logs);
    }
    return logs;
  };
OctoChemConnection.prototype.updateImportationLog =
  async function updateImportationLog(logs) {
    const collection = await this.getImportationLogsCollection();
    logs.dateEnd = Date.now();
    await collection.replaceOne({ _id: logs._id }, logs);
  };

OctoChemConnection.prototype.getImportationLogsCollection =
  async function getCollection() {
    return (await this.getDatabase()).collection('importationLogs');
  };

OctoChemConnection.prototype.getAdminCollection =
  async function getAdminCollection() {
    return this.getCollection('admin');
  };

OctoChemConnection.prototype.setProgress = async function setProgress(
  progress,
) {
  const collection = await this.getAdminCollection();
  await collection.replaceOne({ _id: progress._id }, progress);
};

OctoChemConnection.prototype.getProgress = async function getProgress(
  collectionName,
) {
  const adminCollection = await this.getAdminCollection();
  const _id = `${collectionName}_progress`;
  let progress = await adminCollection.find({ _id }).next();
  if (progress === null) {
    progress = {
      _id,
      state: 'updating',
      seq: 0,
      dateStart: Date.now(),
      dateEnd: 0,
    };
    await adminCollection.insertOne(progress);
  }
  return progress;
};

OctoChemConnection.prototype.getDatabase = async function getDatabase() {
  while (true) {
    try {
      await this.init();
      break;
    } catch (e) {
      debug('Connection to mongo failed, waiting 5s');
      debug(e.stack);
      await delay(5000);
    }
  }
  return this.connection.db(process.env.MONGO_DB_NAME);
};

OctoChemConnection.prototype.init = async function init() {
  if (this.connection) return;

  debug(`Trying to connect to: ${process.env.MONGODB_URL}`);

  this.connection = await this.client.connect();
  debug('Got DB connection');
};
