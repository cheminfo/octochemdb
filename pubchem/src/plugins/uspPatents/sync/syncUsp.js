import Debug from '../../../utils/Debug.js';

import { getFilesToImportForUsp } from './utils/getFilesToImportForUsp.js';
import { syncAllUspFolders } from './utils/http/syncAllUspFolders.js';
import { importUspFiles } from './utils/importUspFiles.js';
/**
 * @description performs the importation of usp patents
 * @param {*} connection - mongo connection
 * @returns {Promise} usp collection
 */
export async function sync(connection) {
  const debug = Debug('syncUsp');
  try {
    // get progress
    const progress = await connection.getProgress('uspPatents');

    // get all files to import
    const allFiles = await syncAllUspFolders(connection);
    debug(allFiles);
    const { files, lastDocument } = await getFilesToImportForUsp(
      connection,
      progress,
      allFiles,
    );
    debug(files);
    //set progress to updating
    progress.state = 'updating';
    await connection.setProgress(progress);
    await importUspFiles(connection, progress, files, { lastDocument });
    // set progress to updated
    progress.state = 'updated';
    await connection.setProgress(progress);
    // create indexes
    const collection = await connection.getCollection('uspPatents');
    await collection.createIndexes([
      { 'data.title': 1 },
      { 'data.language': 1 },
      { 'data.patentNumber': 1 },
      { 'data.applicationType': 1 },
      { 'data.pubchemPatentId': 1 },
      { _id: 1 },
    ]);
  } catch (e) {
    if (connection) {
      debug(e.message, {
        collection: 'uspPatents',
        connection,
        stack: e.stack,
      });
    }
  }
}
