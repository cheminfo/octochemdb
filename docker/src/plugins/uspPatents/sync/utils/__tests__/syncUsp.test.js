import { describe, it, expect } from 'vitest';

import { OctoChemConnection } from '../../../../../utils/OctoChemConnection.js';
import { sync } from '../../syncUsp';

describe('USP Patents', () => {
  it('syncUsp', async () => {
    const connection = new OctoChemConnection();

    await sync(connection);
    const collection = await connection.getCollection('uspPatents');
    const collectionEntry = await collection
      .find({ _id: 'US20230012473A1' })
      .limit(1);
    const result = await collectionEntry.next();
    if (result?._seq) {
      delete result._seq;
    }
    expect(result).toMatchSnapshot();
    await connection.close();
  });
}, 30000);
