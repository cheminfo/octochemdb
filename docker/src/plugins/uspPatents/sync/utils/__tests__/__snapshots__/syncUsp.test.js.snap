// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`USP Patents > syncUsp 1`] = `
{
  "_id": "US-20230012473-A1",
  "data": {
    "abstract": "An airbag deviceis an airbag device for a flying object including an upper cover memberand a lower cover memberas a housing, and a batteryprovided inside the housing and includes an expandable memberwhich is provided inside the housing of the flying objectand is expandable inside the housing, a gas generatorwhich is connected to the expandable memberand receives a predetermined amount of current to make gas flow into the expandable member, and a current supply section which supplies current to the gas generatoraccording to collision of the flying object",
    "dateProduced": 20221231,
    "datePublished": 20230119,
    "patentNumber": 20230012473,
    "pubchemPatentId": "US20230012473A1",
    "title": "AIRBAG DEVICE FOR AIRCRAFT",
  },
}
`;
